### PR TITLE
Use /bin/bash for root shell and suppress some warnings from sysusers

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Comment by mrc0mmand** [comment]
```
exec-specifier.service: Child 8037 belongs to exec-specifier.service. 20:47
exec-specifier.service: Main process exited, code=exited, status=0/SUCCESS (success) 20:47
exec-specifier.service: Running next main command for state start. 20:47
exec-specifier.service: Will spawn child (service_run_next_main): sh 20:47
exec-specifier.service: Passing 0 fds to service 20:47
exec-specifier.service: About to execute sh -c "test /bin/bash = /bin/sh" 20:47
exec-specifier.service: Forked sh as 8038 20:47
Received SIGCHLD from PID 8038 (sh). 20:47
Child 8038 (sh) died (code=exited, status=1/FAILURE) 20:47
exec-specifier.service: Child 8038 belongs to exec-specifier.service. 20:47
exec-specifier.service: Main process exited, code=exited, status=1/FAILURE 20:47
exec-specifier.service: Failed with result 'exit-code'. 20:47
exec-specifier.service: Service will not restart (restart setting) 20:47
exec-specifier.service: Changed start -> failed 20:47
exec-specifier.service: Unit entered failed state. 20:47
src/test/test-execute.c:1080:test_exec_specifier: exec-specifier.service: can_unshare=yes: exit status 1, expected 0
```

I suspect this needs changes here:

https://github.com/systemd/systemd/blob/d8776eed9b6628ccb7d5dad2e26e7405deac9449/test/test-execute/exec-specifier.service#L29

and maybe here as well:

https://github.com/systemd/systemd/blob/d8776eed9b6628ccb7d5dad2e26e7405deac9449/man/systemd.unit.xml#L2119-L2121

**Comment by keszybz** [comment]
Thanks. I adjusted the tests to (hopefully) pass.
I also added one more commit for another issue.

**Review by bluca — APPROVED** [review]

File: src/sysusers/sysusers.c [hunk 1]
```diff
@@ -1478,8 +1486,19 @@ static bool item_equal(Item *a, Item *b) {
...
-                return false;
+        const char *a_shell = pick_shell(a),
+                   *b_shell = pick_shell(b);
+        if (!path_equal_ptr(a_shell, b_shell)) {
+                /* Let's check if the two paths refer to the same file.
+                 * For example, /sbin/nologin and /usr/sbin/nologin.
+                 * If we can't resolve something, treat different paths as different. */
+
+                const char *pa = prefix_roota(arg_root, a_shell),
+                           *pb = prefix_roota(arg_root, b_shell);
```
Comment (poettering) [inline src/sysusers/sysusers.c]: for new code we really should use `chase_symlinks()` everywhere.

Specifically, use chase_symlinks_and_stat() here on the two paths, and then stat_inode_same().

Comment (poettering) [reply src/sysusers/sysusers.c]: that said, I wonder if it wouldn't be better to ignore the status on disk. If this is really just about nologin, then maybe add an explicit check here somewhere:

```c
if (!path_equal_ptr(a_shell, b_shell) &&
    !(is_nologin_shell(a_shell) && is_nologin_shell(b_shell)) 
        return false;
```

benefit: /bin/true and /bin/false would be identified correctly as being equivalent, too.

File: src/sysusers/sysusers.c [hunk 2]
```diff
@@ -1525,22 +1525,22 @@ static int parse_line(const char *fname, unsigned line, const char *buffer) {
         r = extract_many_words(&p, NULL, EXTRACT_UNQUOTE,
                                &action, &name, &id, &description, &home, &shell, NULL);
         if (r < 0)
-                return log_error_errno(r, "[%s:%u] Syntax error.", fname, line);
+                return log_error_errno(r, "%s:%u: Syntax error.", fname, line);
```
Comment (poettering) [inline src/sysusers/sysusers.c]: maybe switch to log_syntax() anyway? would mean clickable links and stuff...

File: src/basic/user-util.c [hunk 3]
```diff
@@ -154,6 +153,17 @@ bool is_nologin_shell(const char *shell) {
                            "/usr/bin/true");
 }
 
+const char* default_root_shell(const char *root) {
+        const char *p = prefix_roota(root, DEFAULT_USER_SHELL);
+
+        /* We want to use the preferred shell, i.e. DEFAULT_USER_SHELL, which usually
+         * will be /bin/bash. But let's fall back to /bin/sh. */
+        if (access(p, X_OK) == 0)
```
Comment (poettering) [inline src/basic/user-util.c]: chase_symlinks() as usual...

(maybe add chase_symlinks_access() that combines chase_symlinks() and then uses faccess() on the final inode)

prefix_roota() should be considered obsolete I guess. We should remove it sooner or later and just use chase_symlinks() everywhere

File: src/nss-systemd/nss-systemd.c [hunk 4]
```diff
@@ -142,24 +142,26 @@ NSS_INITGROUPS_PROTOTYPE(systemd);
 static enum nss_status copy_synthesized_passwd(
                 struct passwd *dest,
                 const struct passwd *src,
+                const char *fallback_shell,
                 char *buffer, size_t buflen,
                 int *errnop) {
 
         size_t required;
+        const char *shell = src->pw_shell ?: fallback_shell;
```
Comment (poettering) [inline src/nss-systemd/nss-systemd.c]: you access `src` here, but assert on it only later.

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/sysusers/sysusers.c (hunk 2 above)
Comment (keszybz) [reply src/sysusers/sysusers.c]: Hmm, I don't think log_syntax does clickable links. But it's better to use it anyway, I changed the patch now.

File: src/sysusers/sysusers.c [hunk 5]
```diff
@@ -1478,8 +1489,34 @@ static bool item_equal(Item *a, Item *b) {
...
+                   *b_shell = pick_shell(b);
+        if (!path_equal_ptr(a_shell, b_shell) &&
+            !(is_nologin_shell(a_shell) && is_nologin_shell(b_shell))) {
+                _cleanup_free_ char *pa = NULL, *pb = NULL;
+
+                r = chase_symlinks(a_shell, arg_root, CHASE_PREFIX_ROOT | CHASE_NONEXISTENT, &pa, NULL);
+                if (ERRNO_IS_RESOURCE(r))
+                        return log_error_errno(r, "Failed to look up path '%s/%s': %m", arg_root, a_shell);
+                if (r < 0)
+                        return false;
```
Comment (poettering) [inline src/sysusers/sysusers.c]: maybe debug log the precise error code?

File: src/sysusers/sysusers.c [hunk 6]
```diff
@@ -1478,8 +1489,34 @@ static bool item_equal(Item *a, Item *b) {
...
+
+                r = chase_symlinks(b_shell, arg_root, CHASE_PREFIX_ROOT | CHASE_NONEXISTENT, &pb, NULL);
+                if (ERRNO_IS_RESOURCE(r))
+                        return log_error_errno(r, "Failed to look up path '%s/%s': %m", arg_root, b_shell);
+                if (r < 0)
+                        return false;
```
Comment (poettering) [inline src/sysusers/sysusers.c]: ditto

File: src/basic/user-util.c [hunk 7]
```diff
@@ -154,6 +154,17 @@ bool is_nologin_shell(const char *shell) {
                            "/usr/bin/true");
 }
 
+const char* default_root_shell(const char *root) {
+        /* We want to use the preferred shell, i.e. DEFAULT_USER_SHELL, which usually
+         * will be /bin/bash. Fall back to /bin/sh if DEFAULT_USER_SHELL is not found,
+         * or any access errors. */
+
+        if (chase_symlinks(DEFAULT_USER_SHELL, root, CHASE_PREFIX_ROOT, NULL, NULL) > 0)
```
Comment (poettering) [inline src/basic/user-util.c]: maybe debug log if not ENOENT?

File: src/sysusers/sysusers.c (hunk 2 above)
Comment (poettering) [reply src/sysusers/sysusers.c]: if not, then we should add it ;-)

**Comment by keszybz** [comment]
Updated with the additional logging.

**Comment by bluca** [comment]
```
2022-08-23T07:08:44.2517634Z gcc-12 -Ilibudev.a.p -I. -I.. -Isrc/basic -I../src/basic -Isrc/fundamental -I../src/fundamental -Isrc/systemd -I../src/systemd -I../src/libsystemd/sd-bus -I../src/libsystemd/sd-device -I../src/libsystemd/sd-event -I../src/libsystemd/sd-hwdb -I../src/libsystemd/sd-id128 -I../src/libsystemd/sd-journal -I../src/libsystemd/sd-netlink -I../src/libsystemd/sd-network -I../src/libsystemd/sd-resolve -Isrc/shared -I../src/shared -Isrc/libsystemd -I/usr/include/blkid -I/usr/include/libmount -I/usr/include/p11-kit-1 -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Werror -std=gnu11 -O3 -g -Wno-format-signedness -Wno-missing-field-initializers -Wno-unused-parameter -Wdate-time -Wendif-labels -Werror=format=2 -Werror=implicit-function-declaration -Werror=incompatible-pointer-types -Werror=int-conversion -Werror=overflow -Werror=override-init -Werror=return-type -Werror=shift-count-overflow -Werror=shift-overflow=2 -Werror=undef -Wfloat-equal -Wimplicit-fallthrough=5 -Winit-self -Wlogical-op -Wmissing-include-dirs -Wmissing-noreturn -Wnested-externs -Wold-style-definition -Wpointer-arith -Wredundant-decls -Wshadow -Wstrict-aliasing=2 -Wstrict-prototypes -Wsuggest-attribute=noreturn -Wunused-function -Wwrite-strings -Wno-maybe-uninitialized -Wno-unused-result -Werror=missing-declarations -Werror=missing-prototypes -fdiagnostics-show-option -fno-common -fno-strict-aliasing -fstack-protector -fstack-protector-strong -fvisibility=hidden --param=ssp-buffer-size=4 -fno-omit-frame-pointer -Werror=shadow -include config.h -Werror -fPIC -pthread -MD -MQ libudev.a.p/src_basic_user-util.c.o -MF libudev.a.p/src_basic_user-util.c.o.d -o libudev.a.p/src_basic_user-util.c.o -c ../src/basic/user-util.c
2022-08-23T07:08:44.2520842Z In file included from ../src/basic/macro.h:452,
2022-08-23T07:08:44.2521262Z                  from ../src/basic/alloc-util.h:9,
2022-08-23T07:08:44.2521644Z                  from ../src/basic/user-util.c:15:
2022-08-23T07:08:44.2660404Z In function ‘default_root_shell’,
2022-08-23T07:08:44.2661513Z     inlined from ‘synthesize_user_creds’ at ../src/basic/user-util.c:193:34:
2022-08-23T07:08:44.2662115Z ../src/basic/user-util.c:164:36: error: ‘%s’ directive argument is null [-Werror=format-overflow=]
2022-08-23T07:08:44.2663000Z   164 |                 log_debug_errno(r, "Failed to look up shell '%s/%s': %m", root, DEFAULT_USER_SHELL);
2022-08-23T07:08:44.2663397Z       |                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2022-08-23T07:08:44.2663879Z ../src/basic/log.h:209:86: note: in definition of macro ‘log_full_errno_zerook’
2022-08-23T07:08:44.2664305Z   209 |                         ? log_internal(_level, _e, PROJECT_FILE, __LINE__, __func__, __VA_ARGS__) \
2022-08-23T07:08:44.2664790Z       |                                                                                      ^~~~~~~~~~~
2022-08-23T07:08:44.2665253Z ../src/basic/log.h:245:41: note: in expansion of macro ‘log_full_errno’
2022-08-23T07:08:44.2665665Z   245 | #define log_debug_errno(error, ...)     log_full_errno(LOG_DEBUG,   error, __VA_ARGS__)
2022-08-23T07:08:44.2666028Z       |                                         ^~~~~~~~~~~~~~
2022-08-23T07:08:44.2666489Z ../src/basic/user-util.c:164:17: note: in expansion of macro ‘log_debug_errno’
2022-08-23T07:08:44.2667062Z   164 |                 log_debug_errno(r, "Failed to look up shell '%s/%s': %m", root, DEFAULT_USER_SHELL);
2022-08-23T07:08:44.2667395Z       |                 ^~~~~~~~~~~~~~~
2022-08-23T07:08:44.2667805Z ../src/basic/user-util.c: In function ‘synthesize_user_creds’:
2022-08-23T07:08:44.2668285Z ../src/basic/user-util.c:164:62: note: format string is defined here
2022-08-23T07:08:44.2668828Z   164 |                 log_debug_errno(r, "Failed to look up shell '%s/%s': %m", root, DEFAULT_USER_SHELL);
2022-08-23T07:08:44.2669211Z       |                                                              ^~
2022-08-23T07:08:44.2669559Z In function ‘default_root_shell’,
2022-08-23T07:08:44.2669992Z     inlined from ‘get_shell’ at ../src/basic/user-util.c:631:21:
2022-08-23T07:08:44.2670542Z ../src/basic/user-util.c:164:36: error: ‘%s’ directive argument is null [-Werror=format-overflow=]
2022-08-23T07:08:44.2671155Z   164 |                 log_debug_errno(r, "Failed to look up shell '%s/%s': %m", root, DEFAULT_USER_SHELL);
2022-08-23T07:08:44.2671531Z       |                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2022-08-23T07:08:44.2671977Z ../src/basic/log.h:209:86: note: in definition of macro ‘log_full_errno_zerook’
2022-08-23T07:08:44.2672407Z   209 |                         ? log_internal(_level, _e, PROJECT_FILE, __LINE__, __func__, __VA_ARGS__) \
2022-08-23T07:08:44.2672788Z       |                                                                                      ^~~~~~~~~~~
2022-08-23T07:08:44.2673248Z ../src/basic/log.h:245:41: note: in expansion of macro ‘log_full_errno’
2022-08-23T07:08:44.2673655Z   245 | #define log_debug_errno(error, ...)     log_full_errno(LOG_DEBUG,   error, __VA_ARGS__)
2022-08-23T07:08:44.2674015Z       |                                         ^~~~~~~~~~~~~~
2022-08-23T07:08:44.2674468Z ../src/basic/user-util.c:164:17: note: in expansion of macro ‘log_debug_errno’
2022-08-23T07:08:44.2675028Z   164 |                 log_debug_errno(r, "Failed to look up shell '%s/%s': %m", root, DEFAULT_USER_SHELL);
2022-08-23T07:08:44.2675375Z       |                 ^~~~~~~~~~~~~~~
2022-08-23T07:08:44.2675747Z ../src/basic/user-util.c: In function ‘get_shell’:
2022-08-23T07:08:44.2676210Z ../src/basic/user-util.c:164:62: note: format string is defined here
2022-08-23T07:08:44.2676747Z   164 |                 log_debug_errno(r, "Failed to look up shell '%s/%s': %m", root, DEFAULT_USER_SHELL);
```

**Comment by keszybz** [comment]
Stupid CI is right.

**Pull request merged**